### ISSUE TITLE
Unpatchification details unclear during training

### ISSUE THREAD

**Issue by sayakpaul** [HiDream-ai/HiDream-I1#35]
https://github.com/HiDream-ai/HiDream-I1/blob/0e0e511d96606274d40a9a2fe7bf6f8ba236b5db/hi_diffusers/models/transformers/transformer_hidream_image.py#L326-L328

Could you please elaborate on why the unpatchify operation is done in such a way during training?

We're trying training with the regular Flow-matching objective (pseudo code):

```py
noisy_model_input = (1.0 - sigmas) * noise + sigmas * model_input
model_pred = transformer(
    hidden_states=noisy_model_input,
    encoder_hidden_states=prompt_embeds,
    pooled_embeds=pooled_prompt_embeds,
    timesteps=timesteps,
    img_sizes=img_sizes,
    img_ids=img_ids,
    return_dict=False,
)[0]
target = model_input - noise
loss = mse_loss(model_pred.float(), target.float())
```

But the unpatchify op will not allow it. Do you have any suggestions? 

Cc: @linoytsaban

**Comment by caiqi** [comment]
einops.rearrange shouldn't block gradients. Could you elaborate on how it fails (e.g., error, zero gradient) and what error message appears?

**Comment by sayakpaul** [comment]
Thing is the output shape changes during `training` when this unpatchification block is hit. It's not trivial for us to know how to deal with that. So, we're currently running training with the `else` block. We are still investigating this.

**Comment by xiecon** [comment]
I encountered a similar issue — the `pipeline` and `patchify` handle images with equal and unequal height/width differently, which causes a mismatch between pred and target.
For example, when the input image is square, I get target: (1, 16, 128, 128) vs pred: (1, 16, 4096, 4).
